File: src/18-challenge/solution.ts
export interface ChildData {
	name: string
	address: string
}

export default function findInAgenda(agenda: string, phone: string): ChildData | null {
	const childData: ChildData = { name: '', address: '' }
	const nameRegex: RegExp = new RegExp(/(?<=<)(?:.*)(?=>)/)
	const phoneRegex: RegExp = new RegExp(/(?<=^|\s)\+\d{1,2}(?:-\d*){3}(?=$|\s)/)
	const replaceRegex: RegExp = new RegExp(/<(?:.*)>|(?<=^|\s)\+\d{1,2}(?:-\d*){3}(?=$|\s)/g)

	let counter: number = 0
	const childrenData: string[] = agenda.split('\n')

	for (let i = 0; i < childrenData.length; i++) {
		if (counter > 1) return null

		const data: string = childrenData[i]
		const dataPhone: string = `${data.match(phoneRegex)?.[0]}`
		if (!dataPhone.match(phone)) continue

		const name: string = `${data.match(nameRegex)?.[0]}`
		const address = data.replaceAll(replaceRegex, '').trim()

		childData.name = name
		childData.address = address
		counter++
	}

	return counter === 1 ? childData : null
}

File: src/18-challenge/solution.py
import re
from typing import TypedDict

ChildData = TypedDict(
    "ChildData",
    {
        "name": str,
        "address": str,
    },
)


def find_in_agenda(agenda: str, phone: str) -> ChildData | None:
    child_data: ChildData = {"name": "", "address": ""}

    name_regex: re.Pattern[str] = re.compile(pattern=r"(?<=<)(?:.*)(?=>)")
    phone_regex: re.Pattern[str] = re.compile(pattern=r"\+\d{1,2}(?:-\d*){3}")

    counter: int = 0

    children_data: list[str] = agenda.split(sep="\n")

    for _, data in enumerate(iterable=children_data):
        if counter > 1:
            return None

        data_phone: re.Match[str] | None = phone_regex.match(string=data)
        if data_phone is None or not data_phone[0].find(phone):
            continue

        name: re.Match[str] | None = name_regex.match(string=data)

        address: str = re.sub(pattern=name_regex, repl=data, string="")
        address = re.sub(pattern=r"<>", repl=data, string="").strip()
        address = re.sub(pattern=phone_regex, repl=data, string="").strip()

        child_data["name"] = name[0] if name is not None else ""
        child_data["address"] = address
        counter += 1

    return child_data if counter == 1 else None
